ocaml/xapi/xapi_pvs_proxy.ml

```diff
@@ -15,37 +15,57 @@
-				~farm:(Db.PVS_proxy.get_farm ~__context ~self:proxy);
 			let dbg = Context.string_of_task __context in
+			let host = Helpers.get_localhost ~__context in
 			let network = Db.VIF.get_network ~__context ~self:vif in
 			let bridge = Db.Network.get_bridge ~__context ~self:network in
-			let name = proxy_port_name bridge in
-			Network.Net.Bridge.add_port dbg ~bridge ~name ~kind:Network_interface.PVS_proxy ~interfaces:[] ();
-			Db.PVS_proxy.set_currently_attached ~__context ~self:proxy ~value:true
-		with Xapi_pvs_cache.No_cache_sr_available ->
-			D.warn "No PVS cache SR available - starting with proxy unattached"
+			let port_name = Xapi_pvs_farm.proxy_port_name bridge in
+			let farm = Db.PVS_proxy.get_farm ~__context ~self:proxy in
+			let sr, vdi = Xapi_pvs_cache.find_or_create_cache_vdi ~__context ~host ~farm in
+			Network.Net.Bridge.add_port dbg ~bridge ~name:port_name ~kind:Network_interface.PVS_proxy ~interfaces:[] ();
+			Xapi_pvs_farm.update_farm_on_localhost ~__context ~self:farm ~vdi ~starting_proxies:[proxy] ();
+			Db.PVS_proxy.set_currently_attached ~__context ~self:proxy ~value:true;
+			Db.PVS_proxy.set_cache_SR ~__context ~self:proxy ~value:sr
+		with e ->
+			let reason =
+				match e with
+				| Xapi_pvs_cache.No_cache_sr_available -> "no PVS cache SR available"
+				| Network_interface.PVS_proxy_connection_error -> "unable to connect to PVS proxy daemon"
+				| _ -> "unknown error"
+			in
+			warn "Unable to enable PVS proxy for VIF %s: %s. Continuing with proxy unattached." (Ref.string_of vif) reason
 	end
 
 let stop ~__context vif proxy =
 	if Db.PVS_proxy.get_currently_attached ~__context ~self:proxy then begin
-		let dbg = Context.string_of_task __context in
-		let network = Db.VIF.get_network ~__context ~self:vif in
-		let bridge = Db.Network.get_bridge ~__context ~self:network in
-		let name = proxy_port_name bridge in
-		Network.Net.Bridge.remove_port dbg ~bridge ~name;
-		Db.PVS_proxy.set_currently_attached ~__context ~self:proxy ~value:false
+		try
+			let dbg = Context.string_of_task __context in
+			let network = Db.VIF.get_network ~__context ~self:vif in
+			let bridge = Db.Network.get_bridge ~__context ~self:network in
+			let port_name = Xapi_pvs_farm.proxy_port_name bridge in
+			let farm = Db.PVS_proxy.get_farm ~__context ~self:proxy in
+			let sr = Db.PVS_proxy.get_cache_SR ~__context ~self:proxy in
+			let vdi = Xapi_pvs_cache.find_cache_vdi ~__context ~sr in
+			Xapi_pvs_farm.update_farm_on_localhost ~__context ~self:farm ~vdi ~stopping_proxies:[proxy] ();
+			Network.Net.Bridge.remove_port dbg ~bridge ~name:port_name;
+			Db.PVS_proxy.set_currently_attached ~__context ~self:proxy ~value:false;
+			Db.PVS_proxy.set_cache_SR ~__context ~self:proxy ~value:Ref.null
+		with e ->
+			let reason =
+				match e with
+				| Xapi_pvs_cache.No_cache_vdi_present -> "no PVS cache VDI found"
+				| Network_interface.PVS_proxy_connection_error -> "unable to connect to PVS proxy daemon"
+				| _ -> "unknown error"
```

Comment:
you could use `Printf.sprintf "unknown error (%s)" __LOC__` to include the location